src/pykeen/nn/init.py

```diff
@@ -308,6 +312,88 @@ def from_triples_factory(
+
+    .. seealso::
+        https://arxiv.org/abs/2110.07875
+    """
+
+    def __init__(
+        self,
+        *,
+        triples_factory: Optional[CoreTriplesFactory] = None,
+        mapped_triples: Optional[MappedTriples] = None,
+        edge_index: Optional[torch.Tensor] = None,
+        dim: int,
+        num_entities: Optional[int] = None,
+        space_dim: int = 0,
+        skip_first_power: bool = True,
+    ) -> None:
+        """
+        Initialize the positional encoding.
+
+        One of `triples_factory`, `mapped_triples` or `edge_index` will be used.
+        The preference order is:
+
+        1. `triples_factory`
+        2. `mapped_triples`
+        3. `edge_index`
+
+        :param triples_factory:
+            the triples factory
+        :param mapped_triples: shape: `(m, 3)`
+            the mapped triples
+        :param edge_index: shape: `(2, m)`
+            the edge index
+        :param dim:
+            the dimensionality
+        :param num_entities:
+            the number of entities. If `None`, it will be inferred from `edge_index`
+        :param space_dim:
+            estimated dimensionality of the space. Used to
+            correct the random-walk diagonal by a factor `k^(space_dim/2)`.
+            In euclidean space, this correction means that the height of
+            the gaussian distribution stays almost constant across the number of
+            steps, if `space_dim` is the dimension of the euclidean space.
+        :param skip_first_power:
+            in most cases the adjacencies diagonal values will be zeros (since reflexive edges are not that common).
+            This flag enables skipping the first matrix power.
+        """
+        if triples_factory is not None:
+            if mapped_triples is not None:
+                logger.warning("Ignoring mapped_triples, since triples_factory is present.")
+            mapped_triples = triples_factory.mapped_triples
+        if mapped_triples is not None:
+            if edge_index is not None:
+                logger.warning("Ignoring edge_index, since mapped_triples is present.")
+            edge_index = mapped_triples[:, 0::2].t()
+        # create random walk matrix
+        adj = torch_ppr.utils.prepare_page_rank_adjacency(edge_index=edge_index, num_nodes=num_entities)
+        num_entities = adj.shape[0]
+        # create a generator of matrix powers
+        powers = iter_matrix_power(matrix=adj, max_iter=dim)
+        # in most cases A**1 diagonal values will be zeros (since reflexive edges are not that common)
+        # so we might want to skip the first
+        if skip_first_power:
+            powers = itertools.islice(powers, 1, None)
+        # prepare an extractor for diagonal entries with densification if necessary
+        diag_extractor = functools.partial(extract_diagonal_sparse_or_dense, eye=sparse_eye(n=num_entities))
+        # an iterator over the power diagonal
+        diagonals = map(diag_extractor, powers)
```

Comment:
Does torch not have some kind of vectorize function for this

Reply:
`powers` is a generator; we do not want to materialize all values (since each of them is a potentially large matrix).